### PR TITLE
JS initializers 8.0

### LINKED ISSUE #30918: JS initializers 8.0
### Description

Per https://github.com/dotnet/aspnetcore/issues/50187 and https://github.com/dotnet/aspnetcore/pull/50860.

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/startup?view=aspnetcore-7.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/fundamentals/startup.md

### Document ID

11f8db3d-9988-e0a1-89c5-c3552ee47b93

### Article author

guardrex

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #30918
Addresses #28161

Javier, Mackinnon ... Guessing on a couple of aspects, particularly the arguments (`options`, `extensions`, and `blazor` of the legacy initializers for the new ones). Also, I'm sure you'll want to update/add remarks on behaviors.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/fundamentals/startup.md](https://github.com/dotnet/AspNetCore.Docs/blob/12a8f9b5480c9d7df780d8df2d4825ef290cfc75/aspnetcore/blazor/fundamentals/startup.md) | [ASP.NET Core Blazor startup](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/startup?branch=pr-en-us-30920) |
| [aspnetcore/blazor/host-and-deploy/webassembly-deployment-layout.md](https://github.com/dotnet/AspNetCore.Docs/blob/12a8f9b5480c9d7df780d8df2d4825ef290cfc75/aspnetcore/blazor/host-and-deploy/webassembly-deployment-layout.md) | [Deployment layout for ASP.NET Core hosted Blazor WebAssembly apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly-deployment-layout?branch=pr-en-us-30920) |

### CODE REVIEW

File: aspnetcore/blazor/fundamentals/startup.md [hunk 1]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+
+<!-- UPDATE 8.0 Remove NOTE -->
+
+> [!NOTE]
+> JS initializers for Blazor Web Apps is an upcoming feature that will appear with the release of .NET 8 in mid-November.
+
+For Blazor Web Apps:
+
+* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeStart` is used to customize the loading process, logging level, and other options. Receives the Blazor options (`options`).
+* `afterWebStarted(blazor)`: Called after the Blazor Web App is ready to receive calls from JS. For example, `afterWebStarted` is used to initialize libraries by making JS interop calls and registering custom elements. The Blazor instance is passed to `afterWebStarted` as an argument (`blazor`).
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:119]: JS interop can't yet be performed at this stage. That's only possible after an interactive Server or WebAssembly runtime starts up.

File: aspnetcore/blazor/fundamentals/startup.md [hunk 2]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+:::moniker range=">= aspnetcore-8.0"
+
+<!-- UPDATE 8.0 Remove NOTE -->
+
+> [!NOTE]
+> JS initializers for Blazor Web Apps is an upcoming feature that will appear with the release of .NET 8 in mid-November.
+
+For Blazor Web Apps:
+
+* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeStart` is used to customize the loading process, logging level, and other options. Receives the Blazor options (`options`).
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:118]: Should this say:
```suggestion
* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeStart` is used to customize the loading process, logging level, and other options. Receives the Blazor Web options (`options`).
```
?

File: aspnetcore/blazor/fundamentals/startup.md [hunk 3]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+* `afterWebStarted(blazor)`: Called after the Blazor Web App is ready to receive calls from JS. For example, `afterWebStarted` is used to initialize libraries by making JS interop calls and registering custom elements. The Blazor instance is passed to `afterWebStarted` as an argument (`blazor`).
+* `beforeServerStart(options, extensions)`: Called before each interactive Server runtime is started. Receives SignalR circuit start options (`options`) and any extensions (`extensions`) added during publishing.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:120]:
```suggestion
* `beforeServerStart(options, extensions)`: Called before the first Server runtime is started. Receives SignalR circuit start options (`options`) and any extensions (`extensions`) added during publishing.
```

The Blazor Server initializer only gets invoked the first time a circuit is created.

File: aspnetcore/blazor/fundamentals/startup.md [hunk 4]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+* `afterServerStarted(blazor)`: Called after each interactive Server runtime is started.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:121]:
```suggestion
* `afterServerStarted(blazor)`: Called after the first interactive Server runtime is started.
```

File: aspnetcore/blazor/fundamentals/startup.md [hunk 5]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+* `beforeWebAssemblyStart(options, extensions)`: Called before each interactive WebAssembly runtime is started. Receives the Blazor options (`options`) and any extensions (`extensions`) added during publishing. For example, options can specify the use of a custom [boot resource loader](xref:blazor/fundamentals/startup#load-client-side-boot-resources).
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:122]: The WebAssembly runtime doesn't stop after it starts, so we can use phrasing that implies there's only ever one WebAssembly runtime:
```suggestion
* `beforeWebAssemblyStart(options, extensions)`: Called before the interactive WebAssembly runtime is started. Receives the Blazor options (`options`) and any extensions (`extensions`) added during publishing. For example, options can specify the use of a custom [boot resource loader](xref:blazor/fundamentals/startup#load-client-side-boot-resources).
```

File: aspnetcore/blazor/fundamentals/startup.md [hunk 6]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+* `afterWebAssemblyStarted(blazor)`: Called after each interactive WebAssembly runtime is started.
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:123]:
```suggestion
* `afterWebAssemblyStarted(blazor)`: Called after the interactive WebAssembly runtime is started.
```

File: aspnetcore/blazor/fundamentals/startup.md [hunk 7]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
 
 ```javascript
-export function beforeStart(options, extensions) {
+export function beforeStart() {
   var customScript = document.createElement('script');
   customScript.setAttribute('src', 'beforeStartScripts.js');
   document.head.appendChild(customScript);
 }
 
-export function afterStarted(blazor) {
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:127]: What's the reason for removing these parameters?

Comment (guardrex) [reply aspnetcore/blazor/fundamentals/startup.md:127]: I'll reverse it ... they just weren't used in the example. Np ... I'll put them back.

File: aspnetcore/blazor/fundamentals/startup.md [hunk 8]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+
+:::moniker-end
+
+:::moniker range=">= aspnetcore-6.0"
+
 The following example demonstrates JS initializers that load custom scripts before and after Blazor has started by appending them to the `<head>` in `beforeStart` and `afterStarted`:
 
 ```javascript
-export function beforeStart(options, extensions) {
+export function beforeStart() {
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/startup.md:189]:
```suggestion
export function beforeStart(options, extensions) {
```

File: aspnetcore/blazor/fundamentals/startup.md [hunk 9]
```diff
@@ -104,27 +104,95 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
   var customScript = document.createElement('script');
   customScript.setAttribute('src', 'beforeStartScripts.js');
   document.head.appendChild(customScript);
 }
 
-export function afterStarted(blazor) {
+export function afterStarted() {
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/startup.md:195]:
```suggestion
export function afterStarted(blazor) {
```

File: aspnetcore/blazor/fundamentals/startup.md (hunk 1 above)
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/startup.md:119]: What should the description be?

File: aspnetcore/blazor/fundamentals/startup.md [hunk 10]
```diff
@@ -104,17 +104,85 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+
+<!-- UPDATE 8.0 Remove NOTE -->
+
+> [!NOTE]
+> JS initializers for Blazor Web Apps is an upcoming feature that will appear with the release of .NET 8 in mid-November.
+
+For Blazor Web Apps:
+
+* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeStart` is used to customize the loading process, logging level, and other options. Receives the Blazor Web options (`options`).
+* `afterWebStarted(blazor)`: Called after the Blazor Web App is ready to receive calls from JS. For example, `afterWebStarted` is used to initialize libraries by making JS interop calls and registering custom elements. The Blazor instance is passed to `afterWebStarted` as an argument (`blazor`).
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/startup.md:119]: How about ...

```suggestion
* `afterWebStarted(blazor)`: Called after the Blazor Web App has started. For example, `afterWebStarted` can be used to register custom elements. The Blazor instance is passed to `afterWebStarted` as an argument (`blazor`). Note that JS interop can't be performed at this stage. JS interop, such as initializing libraries, is only possible after an interactive Server or WebAssembly runtime starts.
```

File: aspnetcore/blazor/fundamentals/startup.md (hunk 1 above)
Comment (MackinnonBuck) [reply aspnetcore/blazor/fundamentals/startup.md:119]: Maybe something like, "Called after all `beforeWebStart` promises resolve. For example, `afterWebStarted` can be used to register Blazor event listeners and custom event types".

File: aspnetcore/blazor/fundamentals/startup.md [hunk 11]
```diff
@@ -104,17 +104,85 @@ To define a JS initializer, add a JS module to the project named `{NAME}.lib.mod
...
+:::moniker range=">= aspnetcore-8.0"
+
+<!-- UPDATE 8.0 Remove NOTE -->
+
+> [!NOTE]
+> JS initializers for Blazor Web Apps is an upcoming feature that will appear with the release of .NET 8 in mid-November.
+
+For Blazor Web Apps:
+
+* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeStart` is used to customize the loading process, logging level, and other options. Receives the Blazor Web options (`options`).
```
Comment (MackinnonBuck) [inline aspnetcore/blazor/fundamentals/startup.md:118]:
```suggestion
* `beforeWebStart(options)`: Called before the Blazor Web App starts. For example, `beforeWebStart` is used to customize the loading process, logging level, and other options. Receives the Blazor Web options (`options`).
```

File: aspnetcore/blazor/fundamentals/startup.md (hunk 10 above)
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/startup.md:119]:
```suggestion
* `afterWebStarted(blazor)`: Called after all `beforeWebStart` promises resolve. For example, `afterWebStarted` can be used to register Blazor event listeners and custom event types. The Blazor instance is passed to `afterWebStarted` as an argument (`blazor`).
```

**Pull request merged**